src/shim/ipc.cc

```diff
@@ -5,20 +5,20 @@
 #include <stddef.h>
 #include <string.h>
 
-#include "shim/gate.h"
+#include "shim/binary_spinning_sem.h"
 
 struct IPCData {
     ShimEvent plugin_to_shadow, shadow_to_plugin;
-    Gate xfer_ctrl_to_plugin, xfer_ctrl_to_shadow;
+    BinarySpinningSem xfer_ctrl_to_plugin, xfer_ctrl_to_shadow;
 };
 
 extern "C" {
 
 void ipcData_init(IPCData *ipc_data) {
     memset(ipc_data, 0, sizeof(IPCData));
```

Comment:
As before, memset is a bit fragile here. Now I remember that this is especially the case for a C++ object, since a memset won't run the constructors of any member objects, if applicable.

As far as I can tell struct init *does* technically work here, but actually is also fragile due to c++ *destuctors*: if IPCData or any of its members have destructors, they'd run on the garbage data at `ipc_data` before calling IPCData's constructor (you haven't defined one, but IIRC there's implicitly one that default-initializes all members).

I think the idiomatic, safe way to do this is to replace `IPCData::init` with a constructor `IPDCata::IPCData`, and here use a placement new.

Demo: https://godbolt.org/z/df9Gjc

Reply:
Alternatively, we could add a static assertion that IPCData (and therefore its members) doesn't have any nontrivial constructor or destructors, in which case the struct-init, or even the memset, should be safe: https://godbolt.org/z/c6nsTa

```
    void ipcData_init(IPCData *ipc_data) {
        static_assert(std::is_trivial<IPCData>());
        static_assert(std::is_trivially_destructible<IPCData>());
        *ipc_data = {};
    }
```

Reply:
Good catch! I think memsetting a class has bitten me before. I had heard of placement `new`, but had forgotten its use since. Fixed!